helix-core/src/main/java/org/apache/helix/controller/stages/TaskGarbageCollectionStage.java

```diff
@@ -81,8 +81,8 @@ public void process(ClusterEvent event) throws Exception {
         if (nextPurgeTime <= currentTime) {
           nextPurgeTime = currentTime + purgeInterval;
           // Find jobs that are ready to be purged
-          Set<String> expiredJobs =
-              TaskUtil.getExpiredJobsFromCache(dataProvider, workflowConfig, workflowContext);
+          Set<String> expiredJobs = TaskUtil
+              .getExpiredJobsFromCache(dataProvider, workflowConfig, workflowContext, manager);
```

Comment:
This is public API change. I know it is added recently, but are we sure this won't break some other code using TaskUtil?

The other way can be simply add a check after line 85 to see if any expiredJobs are in the config and remove them from the expiredJob set.

Reply:
I see the concern here, but doing it in the way you described is not so clean. We need to check for the case of "JobConfig missing from cache but exist in ZK" for all the expiredJobs, and for each of those cases, we need to apply the "isJobExpired" logic again. There is a lot of logic redundancy here. 

Is that better than changing the signature? I think both ways have minor drawbacks, but I personally think changing the signature is fine.

Reply:
Good point. Maybe let us create something like `TaskUtil.getExpiredJobsFromCacheAndCheckZk` which uses `TaskUtil.getExpiredJobsFromCache()`

Reply:
Is that going to be a public function? If so, can we remove that public function after we fix phantom write? 

I don't think this is a big problem; this function is too new to cause large impact imo.

Reply:
We can make something like "TaskUtil.getExpiredJobsFromCacheAndCheckZk" deprecated once we fixed root of the selective update issue. The fix may take several quarters though.

If you believe `TaskUtil.getExpiredJobsFromCache()` is not used by others. Then shall we run this diff against TMC build to see parterners in upstream can build against this one successfully? It would be much more costly if people later find this thing does not work in release time.

Reply:
I think this may be an overkill. I did a code search and didn't find any code calling this function. 

In addition, I can allow `manager` to be null for maximum backward compatibility.

Reply:
Some of the partners code are actually also hosted in github. Code search won't really yield the current usage. In fact last time we  found some issue when integrating with Gobblin. Let me slack you the command to check build. 

If the build works, we can just use this diff without any change.

Reply:
The build seems working. let us just keep this change then.